### DISCUSSION TITLE
Nested enums in state break @ObservableState change detection if not consistently marked

### DISCUSSION THREAD

**Discussion by zgotsch** [pointfreeco/swift-composable-architecture#2887] [Q&A]
I just fixed a bug in my app that was caused by inconsistently applying `@ObservableState` to nested enums within my reducer's `State` (which is also marked `ObservableState`). This caused view for my reducer was not re-rendering in response to changes, and was tricky to track down.

Here is a minimal breaking example:
```swift
import ComposableArchitecture
import SwiftUI

@Reducer
struct Test {
  @ObservableState
  struct State {
    var data: OuterEnum = .inner(.off)
    
    @ObservableState
    enum OuterEnum {
      case inner(InnerEnum)
    }

    enum InnerEnum {
      case on
      case off
      
      func toggle() -> Self {
        switch self {
        case .on:
          return .off
        case .off:
          return .on
        }
      }
    }
  }
  
  enum Action {
    case toggle
  }

  var body: some Reducer<State, Action> {
    Reduce { state, action in
      switch action {
      case .toggle:
        if case let .inner(status) = state.data {
          state.data = .inner(status.toggle())
          return .none
        } else {
          fatalError("help")
        }
      }
    }
  }
}

struct TCAView: View {
  var store: StoreOf<Test>
  
  var body: some View {
    if case let .inner(status) = store.data {
      Text(String(describing: status))
      Button("toggle") {
        store.send(.toggle)
      }
    } else {
      Text("no data")
    }
  }
}

#Preview {
  TCAView(store: Store(initialState: Test.State()) {
    Test()
  })
}
```

The view does not re-render when the toggle button is tapped, though you can see that the state is updating (for example, by using `_printChanges`).

There are two ways to fix this:
- remove `@ObservableState` from `OuterEnum`
- add `@ObservableState` to `InnerEnum`

Is this the expected behavior? I would expect that adding `@ObservableState` would make observation more granular, but not to allow updates to be missed. Please let me know if I have misunderstood something or if this is a bug.

**Comment by mbrandonw** [comment]
Hi @zgotsch, this does seem to be an unfortunate gotcha, specifically with observable enums _and_ nested enums. I understand why it's happening, and we thought we even had code in place that would make this work, but now we can't find it. We're going to investigate and will get back to you.

**Comment by zgotsch** [comment]
I think I've found an additional issue when working with observable enums when only the associated data changes within the same enum case. I have filed https://github.com/pointfreeco/swift-composable-architecture/issues/2899.

**Comment by zgotsch** [comment]
Fixed in #2910.
